stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/experimental/ChatSocket.kt

```diff
@@ -0,0 +1,292 @@
+import io.getstream.chat.android.core.internal.coroutines.DispatcherProvider
+import io.getstream.logging.StreamLog
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.flow.launchIn
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.launch
+
+@Suppress("TooManyFunctions", "LongParameterList")
+internal class ChatSocket private constructor(
+    private val apiKey: String,
+    private val wssUrl: String,
+    private val tokenManager: TokenManager,
+    private val socketFactory: SocketFactory,
+    private val coroutineScope: CoroutineScope,
+    private val lifecycleObserver: StreamLifecycleObserver,
+    private val networkStateProvider: NetworkStateProvider,
+) {
+    private var streamWebSocket: StreamWebSocket? = null
+    private val logger = StreamLog.getLogger("Chat:Socket")
+    private var connectionConf: SocketFactory.ConnectionConf? = null
+    private val listeners = mutableSetOf<SocketListener>()
+    private val chatSocketStateService = ChatSocketStateService()
+    private var socketStateObserverJob: Job? = null
+    private val healthMonitor = HealthMonitor(
+        coroutineScope = coroutineScope,
+        checkCallback = { (chatSocketStateService.currentState as? State.Connected)?.event?.let(::sendEvent) },
+        reconnectCallback = { chatSocketStateService.onWebSocketEventLost() }
+    )
+    private val lifecycleHandler = object : LifecycleHandler {
+        override fun resume() { chatSocketStateService.onResume() }
+        override fun stopped() { chatSocketStateService.onStop() }
+    }
+    private val networkStateListener = object : NetworkStateProvider.NetworkStateListener {
+        override fun onConnected() { chatSocketStateService.onNetworkAvailable() }
+        override fun onDisconnected() { chatSocketStateService.onNetworkNotAvailable() }
+    }
+
+    @Suppress("ComplexMethod")
+    private fun observeSocketStateService(): Job {
+        var socketListenerJob: Job? = null
+
+        fun connectUser(connectionConf: SocketFactory.ConnectionConf) {
+            coroutineScope.launch { startObservers() }
+            this.connectionConf = connectionConf
+            socketListenerJob?.cancel()
+            when (networkStateProvider.isConnected()) {
+                true -> {
+                    streamWebSocket = socketFactory.createSocket(connectionConf).apply {
+                        socketListenerJob = listen().onEach {
+                            when (it) {
+                                is StreamWebSocketEvent.Error -> handleError(it.chatError)
+                                is StreamWebSocketEvent.Message -> handleEvent(it.chatEvent)
+                            }
+                        }.launchIn(coroutineScope)
+                    }
+                }
+                false -> chatSocketStateService.onNetworkNotAvailable()
+            }
+        }
+
+        fun reconnect(connectionConf: SocketFactory.ConnectionConf) {
```

Comment:
Those functions are scoped into the `observeSocketStateService()` function to be sure they are only called from inside of this function